### ISSUE TITLE
[C++] Templatized code not found

### ISSUE THREAD

**Issue by JustusAdam** [github/codeql#18122]
This issue is more me double checking that I did not make a mistake here. It appears that code in uninstantiated templates is  inaccessible for analysis.

I’m trying to use CodeQL on a templatized library but it appears that the code within a function template cannot be analyzed?  I understand that if I instantiate the template somewhere I will be able to analyze it but I just wanted to make sure I did not miss a way to analyze the templated code without instantiation.

```cpp
void target()
{
}

template <typename T>
void entry()
{
    target();
}

void entry2()
{
    target();
}
```

This is the query I run 

```ql
import cpp

from Call c
where c.getTarget().getName() = "target"
select c, c.getLocation().getStartLine()
```

This is the output.

```
|       c        | col1 |
+----------------+------+
| call to target |   13 |
```

If the templated code were found I would expect to also see a call in line 8.

CodeQL version: 2.19.3

**Comment by redsun82** [comment]
:wave: @JustusAdam 

No, you are not doing any mistake, but because of technical reasons to do with the very nature of C++ templates we indeed do not carry out any analysis of templatized uninstantiated code with CodeQL.

Templates have extremely lax requirements on their bodies, which makes them land somewhere between lexing and full parsing. The C++ parser is a weird beast that needs some typing information in order to complete parsing: think of the ambiguity of `a<b>c;` that can be interpreted as declaring a variable `c` of type `a<b>` or as the `(a<b)>c` expression, and there are other corner cases like this. In a sense, a template really gets full C++ citizenship once it's instantiated. And because templates may also only make sense on some unspecified shapes of types or values, we can't even really just fake a random instantiation to get some analysis going. Concepts might help, but not by a great extent.

In a nutshell, I wouldn't say analysis of uninstantiated templates is technically impossible, but it is very challenging to say the least. I can suggest writing example code using those templates and carry out analysis through those examples.

**Comment by JustusAdam** [comment]
Ahh, that makes sense. Thanks for the explanation.
